pkg/proxy/usergating/handler.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package usergating
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+
+	"github.com/redhat-cps/front-proxy/pkg/proxy/usergating/ruleset"
+)
+
+type UserGate struct {
+	GetRuleSet ruleset.RuleSetGetter
+}
+
+func WithUserGating(delegate http.Handler, failed http.Handler, getRuleSet ruleset.RuleSetGetter) http.Handler {
+	gate := &UserGate{
+		GetRuleSet: getRuleSet,
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		u, ok := request.UserFrom(ctx)
+		if !ok {
+			delegate.ServeHTTP(w, r)
```

Comment:
Similar to the comment below, do we want this to be allow-by-default or should this return failed if we can't identify the user?

Reply:
Do all authentication methods give a user.Info?

Reply:
From my testing yes, I've tested with token-file, SA token, client-cert and OIDC and they all result in a user.Info containing a username when authentication is successful

Even [Anonymous auth sets a username](https://github.com/kcp-dev/kubernetes/blob/feature-logical-clusters-1.24/staging/src/k8s.io/apiserver/pkg/authentication/request/anonymous/anonymous.go#L37) AFAICS (although we probably don't want to enable that at the proxy)

Reply:
We'll need to test to confirm, but from my SA testing (where I just used the default SA token from the secret in the root workspace), I think we'll need to have a rule like `^system:serviceaccount:.*:*$` or similar